sintr_analysis_server_example/lib/completion_metrics.dart

```diff
@@ -0,0 +1,228 @@
+  if (deliminator == QUOTE) {
+    // Return quoted string
+    ++start;
+    int end = logMessageText.indexOf('"', start);
+    if (end == -1) throw 'expected value for $key in $logMessageText';
+    return logMessageText.substring(start, end);
+  } else if (deliminator == OPEN_BRACE) {
+    // Return JSON map
+    // TODO (danrubel) handle nested and embedded braces
+    int end = logMessageText.indexOf('}');
+    if (end == -1) throw 'expected matching brace for $key in $logMessageText';
+    return logMessageText.substring(start, end + 1);
+  }
+  return null;
+}
+
+/// Process a log message and return a string representing the result
+/// or `null` if no result from the given log entry.
+String _processLogMessage(int time, String msgType, String logMessageText) {
+  if (msgType == 'Req') {
+    String method = _extractJsonValue(logMessageText, 'method');
+    return _processRequest(time, method, logMessageText);
+  } else if (msgType == 'Res') {
+    String requestId = _extractJsonValue(logMessageText, 'id');
+    return _processResponse(time, requestId, logMessageText);
+  } else if (msgType == 'Noti') {
+    String event = _extractJsonValue(logMessageText, 'event');
+    return _processNotification(time, event, logMessageText);
+  } else if (msgType == 'Read') {
+    return null;
+  } else if (msgType == 'Task') {
+    return null;
+  } else if (msgType == 'Log') {
+    return null;
+  } else if (msgType == 'Perf') {
+    return null;
+  } else if (msgType == 'Watch') {
+    return null;
+  }
+  //throw 'unknown msgType $msgType in $logMessageText';
+  return null;
+}
+
+/// Process a log entry representing a notification
+String _processNotification(int time, String event, String logMessageText) {
+  if (event == 'completion.results') {
+    if (!logMessageText.endsWith(',"isLast"::true}}')) return null;
+    var prefix = '{"event"::"completion.results","params"::{"id"::"';
+    if (!logMessageText
+        .startsWith(prefix)) throw 'expected $prefix in $logMessageText';
+    int start = prefix.length;
+    int end = logMessageText.indexOf('"', start);
+    if (end == -1) throw 'expected notification id in $logMessageText';
+    String notificationId = logMessageText.substring(start, end);
+    _Completion completion = _notificationMap.remove(notificationId);
+    if (completion ==
+        null) throw 'expected completion request for $logMessageText';
+    return _composeExtractionResult(completion, time - completion.requestTime);
+  }
+  return null;
+}
+
+/// Process a log entry representing a request
+String _processRequest(int time, String method, String logMessageText) {
+  // Look for completion requests
+  if (method == 'completion.getSuggestions') {
+    String requestId = _extractJsonValue(logMessageText, 'id');
+    _requestMap[requestId] = new _Completion(time);
+  }
+  return null;
+}
+
+/// Process a log entry representing a response
+String _processResponse(int time, String requestId, String logMessageText) {
```

Comment:
The _processLogMessage delegates in a symmetric way so that it does not have to know about delegate internals. It's useful when composing/testing the query extractor and when reusing existing code to build others.